src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -1298,4 +1308,112 @@ protected final void incrementActionsDone()
+     * @param targetPosition the position to work at.
+     * @return BlockPos position to work from.
+     */
+    public BlockPos getWorkingPosition(BlockPos targetPosition)
+    {
+        return targetPosition;
+    }
+
+    /**
+     * Calculates the working position.
+     * <p>
+     * Takes a min distance from width and length.
+     * <p>
+     * Then finds the floor level at that distance and then check if it does contain two air levels.
+     *
+     * @param distance the extra distance to apply away from the building.
+     * @param targetPos the target position which needs to be worked.
+     * @param offset an additional offset
+     * @return BlockPos position to work from.
+     */
+    public BlockPos getWorkingPosition(final int distance, final BlockPos targetPos, final int offset)
+    {
+        if (offset > MAX_ADDITIONAL_RANGE_TO_BUILD)
+        {
+            return targetPos;
+        }
+
+        @NotNull final EnumFacing[] directions = {EnumFacing.EAST, EnumFacing.WEST, EnumFacing.NORTH, EnumFacing.SOUTH};
+
+        //then get a solid place with two air spaces above it in any direction.
+        for (final EnumFacing direction : directions)
+        {
+            @NotNull final BlockPos positionInDirection = getPositionInDirection(direction, distance + offset, targetPos);
+            if (EntityUtils.checkForFreeSpace(world, positionInDirection)
+                    && world.getBlockState(positionInDirection.up()).getBlock() != Blocks.SAPLING)
+            {
+                return positionInDirection;
+            }
+        }
+
+        //if necessary we can could implement calling getWorkingPosition recursively and add some "offset" to the sides.
+        return getWorkingPosition(distance, targetPos, offset + 1);
+    }
+
+    /**
+     * Gets a floorPosition in a particular direction.
+     *
+     * @param facing   the direction.
+     * @param distance the distance.
+     * @param targetPos the position to work at.
+     * @return a BlockPos position.
+     */
+    @NotNull
+    private BlockPos getPositionInDirection(final EnumFacing facing, final int distance, final BlockPos targetPos)
+    {
+        return getFloor(targetPos.offset(facing, distance));
+    }
+
+    /**
+     * Calculates the floor level.
+     *
+     * @param position input position.
+     * @return returns BlockPos position with air above.
+     */
+    @NotNull
+    private BlockPos getFloor(@NotNull BlockPos position)
+    {
+        final BlockPos floor = getFloor(position, 0);
+        if (floor == null)
+        {
+            return position;
+        }
+        return floor;
+    }
+
+    /**
+     * Calculates the floor level.
+     *
+     * @param position input position.
+     * @param depth    the iteration depth.
+     * @return returns BlockPos position with air above.
+     */
+    @Nullable
+    private BlockPos getFloor(@NotNull BlockPos position, int depth)
+    {
+        if (depth > 50)
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Assign this magic number 50 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)